validations/src/validations.rs

```diff
@@ -774,3 +753,89 @@ mod tests {
         assert_eq!(block_reward(1_750_000 * 100), 0);
     }
 }
+
+/// Diffs to apply to an utxo set. This type does not contains a
+/// reference to the original utxo set.
+#[derive(Default)]
+pub struct Diff {
+    utxos_to_add: UnspentOutputsPool,
+    utxos_to_remove: HashSet<OutputPointer>,
+    utxos_to_remove_dr: Vec<OutputPointer>,
+}
+
+impl Diff {
+    pub fn apply(mut self, utxo_set: &mut UnspentOutputsPool) {
+        for (output_pointer, output) in self.utxos_to_add.drain() {
+            utxo_set.insert(output_pointer, output);
+        }
+
+        for output_pointer in self.utxos_to_remove.iter() {
+            utxo_set.remove(output_pointer);
+        }
+
+        for output_pointer in self.utxos_to_remove_dr.iter() {
+            utxo_set.remove(output_pointer);
+        }
+    }
+}
+
+/// Contains a reference to an UnspentOutputsPool plus subsequent
+/// insertions and deletions to performed on that pool.
+/// Use `.take_diff()` to obtain an instance of the `Diff` type.
+pub struct UtxoDiff<'a> {
+    diff: Diff,
+    utxo_pool: &'a UnspentOutputsPool,
+}
+
+impl<'a> UtxoDiff<'a> {
+    /// Create a new UtxoDiff without additional insertions or deletions
+    pub fn new(utxo_pool: &'a UnspentOutputsPool) -> Self {
+        UtxoDiff {
+            utxo_pool,
+            diff: Default::default(),
+        }
+    }
+
+    /// Record an insertion to perform on the utxo set
+    pub fn insert_utxo(&mut self, output_pointer: OutputPointer, output: Output) {
+        self.diff.utxos_to_add.insert(output_pointer, output);
+    }
+
+    /// Record a deletion to perform on the utxo set
+    pub fn remove_utxo(&mut self, output_pointer: OutputPointer) {
+        match self.diff.utxos_to_add.remove(&output_pointer) {
```

Comment:
clippy doesnt like it

```suggestion
        if let None = self.diff.utxos_to_add.remove(&output_pointer) {
   self.diff.utxos_to_remove.insert(output_pointer);
 }
```